src/guid.rs

```diff
@@ -0,0 +1,123 @@
+use std::{fmt, str, ops};
+
+// TODO: We could store a 13th b'\0' byte for a C-compatible null terminator. Is
+// this worth doing?
+#[derive(Clone, Copy, Eq, PartialEq, Ord, PartialOrd, Hash)]
+pub struct Guid([u8; 12]);
+
+// These are here so that we can create these as const/static (without needing lazy_static)
+pub(crate) const ROOT_GUID: Guid = Guid(*b"root________");
+
+pub(crate) const USER_CONTENT_ROOTS: [Guid; 4] = [
+    Guid(*b"toolbar_____"),
+    Guid(*b"menu________"),
+    Guid(*b"unfiled_____"),
+    Guid(*b"mobile______"),
+];
+
+impl Guid {
+    #[inline]
+    pub fn try_new(s: &str) -> Option<Guid> {
+        if !Guid::is_valid(s) {
+            None
+        } else {
+            assert!(s.is_char_boundary(12));
+            let mut guid = Guid([0u8; 12]);
+            guid.0.copy_from_slice(s.as_bytes());
+            Some(guid)
+        }
+    }
+
+
+    #[inline]
+    pub fn new(s: &str) -> Guid {
+        match Guid::try_new(s) {
+            Some(g) => g,
+            // We don't get to format the string passed to e.g. `expect`.
+            None => panic!("Invalid guid: {:?}", s)
+        }
+    }
+
+    #[inline]
+    pub fn as_bytes(&self) -> &[u8] {
+        &self.0
+    }
+
+    #[inline]
+    pub fn as_str(&self) -> &str {
+        // We actually could use from_utf8_unchecked here, and depending on how
+        // often we end up doing this, it's arguable that we should. We know
+        // already this is valid utf8, since we know that we only ever create
+        // these while respecting is_valid (and moreover, we assert that
+        // `s.is_char_boundary(12)` above).
+        str::from_utf8(self.as_bytes()).unwrap()
+    }
+
+    /// Equivalent to `PlacesUtils.isValidGuid`.
+    #[inline]
+    pub fn is_valid(s: &str) -> bool {
```

Comment:
In theory, we shouldn't have invalid GUIDs on the client or server, but, in practice, we've seen issues with both. I wonder if it makes sense to represent `Guid` as an enum, or some other kind of opaque type. We could use your fast, compact representation for valid GUIDs, and owned `String`s for invalid ones. Or maybe that's not worth the complexity, and we should handle invalid GUIDs one level up. WDYT?